File: app/src/main/java/org/bluechat/blueninemenmoris/model/Player.java
package org.bluechat.blueninemenmoris.model;


public abstract class Player {
	
	protected String name;
	protected int gamesWon;
	protected int numPieces;
	protected int numPiecesOnBoard;
	protected Token playerToken;
	protected boolean canFly;
	protected Actor[] actors;
	
	protected Player() {
		gamesWon = 0;
		numPiecesOnBoard = 0;
		canFly = false;
	}
	
	protected Player(Token player, int numPiecesPerPlayer) throws GameException {
		this();
		if(player != Token.PLAYER_1 && player != Token.PLAYER_2) {
			throw new GameException(""+getClass().getName()+" - Invalid Player Token: "+player);
		} else {
			numPieces = numPiecesPerPlayer;
			playerToken = player;
            actors = new Actor[numPieces];
		}
	}
	
	public void reset() {
		numPiecesOnBoard = 0;
		canFly = false;
	}
	
	public String getName() {
		return name;
	}
	
	public void setName(String name) {
		this.name = name;
	}
	
	public int getNumPieces() {
		return numPieces;
	}
	
	public int getNumPiecesOnBoard() {
		return numPiecesOnBoard;
	}
	
	public int getNumPiecesLeftToPlace() {
		return (numPieces - numPiecesOnBoard);
	}
	
	public int raiseNumPiecesOnBoard() {
		canFly = false; // it's still placing pieces
		return ++numPiecesOnBoard;
	}
	
	public int lowerNumPiecesOnBoard() {
		if(--numPiecesOnBoard == 3) {
			canFly = true;
		}
		return numPiecesOnBoard;
	}
	
	public Token getPlayerToken() {
		return playerToken;
	}
	
	public boolean canItFly() {
		return canFly;
	}
	
	public abstract boolean isAI();

    public Actor[] getActors() {
        return actors;
    }
    public void setActors(int sx, int sy){
        for (int i = 0; i < numPieces; i++) {
            actors[i] = new Actor((i + 1)*sx, sy);
        }
    }
}

File: app/src/main/java/org/bluechat/blueninemenmoris/model/Player.kt
package org.bluechat.blueninemenmoris.model


abstract class Player protected constructor() {

    var name: String = ""
    protected var gamesWon: Int = 0
    var numPieces: Int = 0
        protected set
    var numPiecesOnBoard: Int = 0
        protected set
    lateinit var playerToken: Token
        protected set
    protected var canFly: Boolean = false
    lateinit var actors: Array<Actor?>
        protected set

    val numPiecesLeftToPlace: Int
        get() = numPieces - numPiecesOnBoard

    abstract val isAI: Boolean

    init {
        gamesWon = 0
        numPiecesOnBoard = 0
        canFly = false
    }

    @Throws(GameException::class)
    protected constructor(player: Token, numPiecesPerPlayer: Int) : this() {
        if (player != Token.PLAYER_1 && player != Token.PLAYER_2) {
            throw GameException("" + javaClass.name + " - Invalid Player Token: " + player)
        } else {
            numPieces = numPiecesPerPlayer
            playerToken = player
            actors = arrayOfNulls(numPieces)
        }
    }

    fun reset() {
        numPiecesOnBoard = 0
        canFly = false
    }

    fun raiseNumPiecesOnBoard(): Int {
        canFly = false // it's still placing pieces
        return ++numPiecesOnBoard
    }

    fun lowerNumPiecesOnBoard(): Int {
        if (--numPiecesOnBoard == 3) {
            canFly = true
        }
        return numPiecesOnBoard
    }

    fun canItFly(): Boolean {
        return canFly
    }

    fun setActors(sx: Int, sy: Int) {
        for (i in 0 until numPieces) {
            actors[i] = Actor((i + 1) * sx, sy, i + 1)
        }
    }

    fun getActorAt(index: Int): Actor? {
        var act: Actor? = null
        for (actor in actors) {
            if (actor?.placedIndex == index) {
                act = actor
            }
        }
        return act
    }

    fun getActorByNumber(index: Int): Actor? {
        var act: Actor? = null
        for (actor in actors) {
            if (actor?.number == index) {
                act = actor
            }
        }
        return act
    }
}
